utilities/transactions/write_unprepared_txn.h

```diff
@@ -212,7 +212,7 @@ class WriteUnpreparedTxn : public WritePreparedTxn {
   friend class WriteUnpreparedTxnDB;
 
   const std::map<SequenceNumber, size_t>& GetUnpreparedSequenceNumbers();
-  Status WriteRollbackKeys(const TransactionKeyMap& tracked_keys,
+  Status WriteRollbackKeys(const LockTracker& tracked_keys,
                            WriteBatchWithIndex* rollback_batch,
                            ReadCallback* callback, const ReadOptions& roptions);
```

Comment:
Do I understand this code correctly: it walks the tracked_keys to get a set of rowkeys that were modified by the transaction.

I'm wondering if/how this should work with a Range-based Lock Tracker (RBLT)?

Should RBLT be able to return the set of modified rowkeys? (What if it returns a set of ranges that cover all the rowkeys instead? I assume that won't work?)

(I'll need to verify this, but I suspect that RangeLocking doesn't work with WriteUnprepared in the current Range-Locking code... when range_locking=1, `tracked_keys` is empty even if the transaction has made modifications...)

Reply:
Found `WriteUnpreparedTxn::untracked_keys_` ... this looks like it could be used as a workaround...  (it would be a lousy one:  Range Locking is capable of "Lock Escalation" where adjacent locks are merged together and so a big transaction can be run without consuming O(#size-of-changes) RAM... if we just put all modified row keys into untracked_keys_ we'll get the O(#size-of-changes) memory consumption back)

Reply:
@lth , Herman mentioned you could enlighten us here (on the questions in the first comment)?

Reply:
I think it helps to give some context on the goals of write unprepared. The goal of write unprepared is to reduce memory usage by large transactions. Today there are 3 ways memory would grow with size of transaction:
1. write batch
2. tracked keys
3. lock map

The solution to #1 was to flush the write batch into the db periodically and this is done. We still don't have an implemented solution for #2/#3, but the idea is that we could approximate many point locks by merging into a single range lock and save on memory for #2/#3. 

So we would use tracked_keys for a few purposes in write unprepared:
1. to get the list of keys to rollback 
2. to get the list of keys to unlock

I'm guessing the question is how #1 would work in context of range locks. I think there were two ways we were planning to go with 1. reading the list of keys to rollback from the WAL or 2. doing 2 snapshot reads (one at latest, one at the seqno to rollback to) in the tracked ranges to determine the delta in changed keys, and then rolling back the modified keys.

Right now, I think reusing tracked keys would be cleaner than reading from WAL, but this remains an open question. I think this would depend on the performance penalty of using range locking.